internal/cli/atlas/clusters/advancedsettings/update.go

```diff
@@ -0,0 +1,152 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package advancedsettings
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/mongodb/mongodb-atlas-cli/internal/cli"
+	"github.com/mongodb/mongodb-atlas-cli/internal/cli/atlas/commonerrors"
+	"github.com/mongodb/mongodb-atlas-cli/internal/cli/require"
+	"github.com/mongodb/mongodb-atlas-cli/internal/config"
+	"github.com/mongodb/mongodb-atlas-cli/internal/flag"
+	"github.com/mongodb/mongodb-atlas-cli/internal/store"
+	"github.com/mongodb/mongodb-atlas-cli/internal/usage"
+	"github.com/openlyinc/pointy"
+	"github.com/spf13/cobra"
+	atlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+const (
+	updateTmpl                   = "Updating advanced configuration settings of your cluster'.\n"
+	defaultReadConcern           = "available"
+	defaultWriteConcern          = "1"
+	defaultSampleSizeBIConnector = 1000
+)
+
+type UpdateOpts struct {
+	cli.GlobalOpts
+	cli.OutputOpts
+	name                             string
+	defaultReadConcern               string
+	defaultWriteConcern              string
+	minimumEnabledTLSProtocol        string
+	noTableScan                      bool
+	disableFailIndexKeyTooLong       bool
+	disableJavascript                bool
+	oplogMinRetentionHours           float64
+	oplogSizeMB                      int64
+	sampleRefreshIntervalBIConnector int64
+	sampleSizeBIConnector            int64
+	store                            store.AtlasClusterConfigurationOptionsUpdater
+}
+
+func (opts *UpdateOpts) initStore(ctx context.Context) func() error {
+	return func() error {
+		var err error
+		opts.store, err = store.New(store.AuthenticatedPreset(config.Default()), store.WithContext(ctx))
+		return err
+	}
+}
+
+func (opts *UpdateOpts) Run() error {
+	r, err := opts.store.UpdateAtlasClusterConfigurationOptions(opts.ConfigProjectID(), opts.name, opts.newProcessArgs())
+	if err != nil {
+		return commonerrors.Check(err)
+	}
+
+	return opts.Print(r)
+}
+
+func (opts *UpdateOpts) newProcessArgs() *atlas.ProcessArgs {
+	args := &atlas.ProcessArgs{
+		DefaultReadConcern:               opts.defaultReadConcern,
+		DefaultWriteConcern:              opts.defaultWriteConcern,
+		MinimumEnabledTLSProtocol:        opts.minimumEnabledTLSProtocol,
+		SampleSizeBIConnector:            pointy.Int64(opts.sampleSizeBIConnector),
+		SampleRefreshIntervalBIConnector: pointy.Int64(opts.sampleRefreshIntervalBIConnector),
+		NoTableScan:                      pointy.Bool(opts.noTableScan),
+	}
```

Comment:
is this a PATCH or a PUT? you may be over-writing settings here by defining all the options

Reply:
to be clear what happens if I do
```bash
atlas cluster advancedSettings update <clusterName> --projectId <projectId> --writeConcern majority

atlas cluster advancedSettings update <clusterName> --projectId <projectId> --readConcern  majority
```

what's the value of `writeConcern` after the second call?

Reply:
it is a [PATCH](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#operation/updateAdvancedConfigurationOptionsForOneCluster)

Reply:
You are right:
```
./bin/atlas cluster advancedSettings update Cluster87819 --writeConcern majority

~/workspace/mongodb-atlas-cli  CLOUDP-139997 *1 ❯ ./bin/atlas cluster advancedSettings describe Cluster87819 -o json                                                                                                            1.18
{
  "defaultReadConcern": "available",
  "defaultWriteConcern": "majority",
  "minimumEnabledTlsProtocol": "TLS1_2",
  "javascriptEnabled": true,
  "noTableScan": false,
  "sampleSizeBIConnector": 1000,
  "sampleRefreshIntervalBIConnector": 0
}

./bin/atlas cluster advancedSettings update Cluster87819 --readConcern  local                                                                                                 1.18
Updating advanced configuration settings of your cluster'.


./bin/atlas cluster advancedSettings describe Cluster87819 -o json                                                                                                            1.18
{
  "defaultReadConcern": "local",
  "defaultWriteConcern": "1",
  "minimumEnabledTlsProtocol": "TLS1_2",
  "javascriptEnabled": true,
  "noTableScan": false,
  "sampleSizeBIConnector": 1000,
  "sampleRefreshIntervalBIConnector": 0
}
```

Reply:
Updated. Now it is working as expected.

```
./bin/atlas cluster advancedSettings update Cluster87819 --writeConcern majority                                                                                           1.18
Updating advanced configuration settings of your cluster'.
~/workspace/mongodb-atlas-cli  CLOUDP-139997 *1 +1 ❯ ./bin/atlas cluster advancedSettings describe Cluster87819 -o json                                                                                                         1.18
{
  "defaultReadConcern": "local",
  "defaultWriteConcern": "majority",
  "minimumEnabledTlsProtocol": "TLS1_2",
  "javascriptEnabled": true,
  "noTableScan": false,
  "sampleSizeBIConnector": 1000,
  "sampleRefreshIntervalBIConnector": 0
}
~/workspace/mongodb-atlas-cli  CLOUDP-139997 *1 +1 ❯ ./bin/atlas cluster advancedSettings update Cluster87819 --readConcern  local                                                                                              1.18
Updating advanced configuration settings of your cluster'.
~/workspace/mongodb-atlas-cli  CLOUDP-139997 *1 +1 ❯ ./bin/atlas cluster advancedSettings describe Cluster87819 -o json                                                                                                         1.18
{
  "defaultReadConcern": "local",
  "defaultWriteConcern": "majority",
  "minimumEnabledTlsProtocol": "TLS1_2",
  "javascriptEnabled": true,
  "noTableScan": false,
  "sampleSizeBIConnector": 1000,
  "sampleRefreshIntervalBIConnector": 0
}
```